File: PRNumber/app/src/main/java/com/pplus/prnumberuser/apps/setting/data/FaqAdapter.java
package com.pplus.prnumberuser.apps.setting.data;

import android.content.Context;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.pplus.prnumberuser.R;
import com.pplus.prnumberuser.core.network.model.dto.Faq;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by ksh on 2016-09-28.
 */

public class FaqAdapter extends RecyclerView.Adapter<FaqAdapter.ViewHolder>{

    private static final String TAG = FaqAdapter.class.getSimpleName();
    private ArrayList<Faq> mDataList;
    private OnItemClickListener listener;
    private Context mContext;

    public FaqAdapter(Context context){

        mDataList = new ArrayList<>();
        mContext = context;
    }
    public interface OnItemClickListener{

        void onItemClick(int position);
    }

    public Faq getItem(int postion){

        return mDataList != null ? mDataList.get(postion) : null;
    }

    public void setOnItemClickListener(OnItemClickListener listener){

        this.listener = listener;
    }

    public void addAll(List<Faq> data){

        mDataList.addAll(data);
        notifyDataSetChanged();
    }

    public void clear(){

        mDataList.clear();
    }

    @Override
    public int getItemViewType(int position){

        return position;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType){

        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.common_item_two_line, parent, false);
        ViewHolder viewHolder = new ViewHolder(view);

        return viewHolder;
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position){

        final Faq item = mDataList.get(position);

        holder.itemView.setOnClickListener(new View.OnClickListener(){

            @Override
            public void onClick(View view){

                if(listener != null) {
                    listener.onItemClick(holder.getAdapterPosition());
                }
            }
        });

        if(item != null) {

            holder.text_subject.setText(item.getSubject()); // 제목
        }
    }

    @Override
    public int getItemCount(){

        return mDataList != null ? mDataList.size() : null;
    }

    public class ViewHolder extends RecyclerView.ViewHolder{

        public TextView text_subject, text_date;

        public ViewHolder(View itemView){

            super(itemView);

            text_subject = (TextView) itemView.findViewById(R.id.text_subject);
            text_subject.setMaxLines(2);
            text_date = (TextView) itemView.findViewById(R.id.text_date);
            text_date.setVisibility(View.GONE);
        }

    }


}

File: WowBox/app/src/main/java/com/lejel/wowbox/apps/faq/data/FaqAdapter.kt
package com.lejel.wowbox.apps.faq.data

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.lejel.wowbox.R
import com.lejel.wowbox.core.network.model.dto.Faq
import com.lejel.wowbox.databinding.ItemFaqBinding


/**
 * Created by imac on 2018. 1. 8..
 */
class FaqAdapter() : RecyclerView.Adapter<FaqAdapter.ViewHolder>() {

    var mDataList: MutableList<Faq>? = null
    var listener: OnItemClickListener? = null
    var mOpenFaq:Faq? = null

    interface OnItemClickListener {

        fun onItemClick(position: Int)
    }

    init {
        this.mDataList = ArrayList()
    }

    fun setOnItemClickListener(listener: OnItemClickListener) {

        this.listener = listener
    }

    fun getItem(position: Int): Faq {

        return mDataList!![position]
    }

    fun getDataList(): MutableList<Faq>? {

        return mDataList
    }

    fun add(data: Faq) {

        if (mDataList == null) {
            mDataList = ArrayList<Faq>()
        }
        mDataList!!.add(data)
        notifyDataSetChanged()
    }

    fun addAll(dataList: List<Faq>) {

        if (this.mDataList == null) {
            this.mDataList = ArrayList<Faq>()
        }

        this.mDataList!!.addAll(dataList)
        notifyDataSetChanged()
    }

    fun replaceData(position: Int, data: Faq) {
        if (position != -1 && mDataList!!.size > 0) {
            mDataList!!.removeAt(position)
            mDataList!!.add(position, data)
            notifyItemChanged(position)
        }

    }

    fun clear() {
        notifyItemRangeRemoved(0, mDataList!!.size)
        mDataList = ArrayList<Faq>()
    }

    fun setDataList(dataList: MutableList<Faq>) {

        this.mDataList = dataList
        notifyDataSetChanged()
    }

    class ViewHolder(val binding: ItemFaqBinding) : RecyclerView.ViewHolder(binding.root) {

        init {
        }
    }

    override fun getItemCount(): Int {
        return mDataList!!.size
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {

        val item = mDataList!![position]

        holder.binding.textFaqTitle.text = item.title
        holder.binding.textFaqContents.text = item.contents

        if(mOpenFaq != null && mOpenFaq!!.seqNo == item.seqNo){
            holder.binding.textFaqContents.visibility = View.VISIBLE
            holder.binding.imageFaqArrow.setImageResource(R.drawable.ic_faq_arrow_up)
        }else{
            holder.binding.textFaqContents.visibility = View.GONE
            holder.binding.imageFaqArrow.setImageResource(R.drawable.ic_faq_arrow_down)
        }

        holder.binding.layoutFaqTitle.setOnClickListener {
            if(mOpenFaq != null && mOpenFaq!!.seqNo == item.seqNo){
                mOpenFaq = null
            }else{
                mOpenFaq = item
            }

            notifyDataSetChanged()
        }

    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = ItemFaqBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }
}